Add sortText for completion items ordering

heavenshell opened the pull request:
## Summary

Support sortText for ordering.

## Detail

In `deno lsp`, some completion items(folder, file paths) would be convenient by sorted completion items.

Before
<img width="762" alt="スクリーンショット 2021-09-20 1 35 02" src="https://user-images.githubusercontent.com/56591/133935634-751bb6dc-0bb7-4ced-8733-6d023364599a.png">

After
<img width="762" alt="スクリーンショット 2021-09-20 1 33 42" src="https://user-images.githubusercontent.com/56591/133935642-d9db7cba-9cda-4976-85e2-ea6a9efd4c80.png">

Thank you!

mattn commented on autoload/lsp/omni.vim:
```diff
@@ -288,6 +295,11 @@ function! lsp#omni#get_vim_completion_items(options) abort
+    if len(l:items) > 0 && has_key(l:items[0], 'sortText')
+      " If first item contains sortText, maybe we can use sortText
+      call sort(l:items, function('s:sort_by_sorttext'))
+    endif
```
This is my opinion. DO NOT CHANGE CODES AGAINST THIS REVIEW COMMENT PLEASE.

Document says:
> A string that should be used when comparing this item with other items. When falsy the label is used as the sort text for this item.

So the sorting should always be done always like.

```vim
function! s:sort(i1, i2) abort
  let l:text1 = get(a:1, 'sortText', get(a:1, 'label', ''))
  let l:text2 = get(a:2, 'sortText', get(a:2, 'label', ''))
  return l:text1 == l:text2 ? 0 : l:text1 > l:text2 ? 1 : -1
endfunction
```

heavenshell replied:
Thank you for review comment.

I'm worried about breaking existing behavior and performance.
So I intent to sorting only LS supported

@prabirshrestha mention about performance impact I agree that.

what about do someting like this?
User can config sort by each LS server but LS does not support sort, sort by label.

```
" per-server configuration options

config: { 'sort': { 'max': 1000 } }
```

```
function! s:sort(i1, i2) abort
  let l:text1 = get(a:1, 'sortText', get(a:1, 'label', ''))
  let l:text2 = get(a:2, 'sortText', get(a:2, 'label', ''))
  return l:text1 == l:text2 ? 0 : l:text1 > l:text2 ? 1 : -1
endfunction

function! lsp#omni#get_vim_completion_items(options) abort
  " snip

  let l:sort = has_key(l:server_info, 'config') && has_key(l:server_info['config'], 'sort') ? l:server_info['config']['sort'] : { 'max': '1000' }

  if len(l:items) > l:sort['max']
    call sort(l:items, s:sort)
  endif

  " snip
endfunction
```

rhysd replied:
`label` property is not optional.

```vim
function! s:sort(i1, i2) abort
  let l:text1 = get(a:1, 'sortText', a:1['label'])
  let l:text2 = get(a:2, 'sortText', a:2['label'])
  return l:text1 == l:text2 ? 0 : l:text1 > l:text2 ? 1 : -1
endfunction
```

This would slightly be better in terms of performance.

rhysd replied:
And one more point. Falsy values include empty string. When `sortText` property is empty string, `label` should be used instead.

heavenshell replied:
Thank you for comment!

I just update when sortText is empty string, use label instead and get label directory(`a:1['label'])`)

prabirshrestha commented:
This could have a perf impact.  We should pass an option to ignore sorting in `lsp#omni#get_vim_completion_items` options.

Some lang servers return 1000+ items and if folks complain about perf issue we can have a global flag for max number of items before it ignores sort. This flag we can add it later when needed.

rhysd commented:
Should the comparison be case sensitive or case insensitive?

heavenshell commented:
@mattn, @prabirshrestha, @rhysd 

Thank you for reviewing. I updated to use each LS(like filter).
User should set config if they want to use sort.
And if sortText is empty or not set, use label instead.

Could you review this again? 🙏 

>Should the comparison be case sensitive or case insensitive?

How about using `g:lsp_ignorecase`?
I added `g:lsp_ignorecase` flag to use case sensitive or case insensitive in sort function.

heavenshell commented on autoload/lsp/omni.vim:
```diff
@@ -264,6 +264,21 @@ function! s:get_completion_result(server_name, data) abort
+    let l:text2 = get(a:i2, 'sortText')
+
+    " sortText is possibly empty string
+    let l:text1 = l:text1 ? l:text1 : a:i1['label']
+    let l:text2 = l:text2 ? l:text2 : a:i2['label']
```
[memo]
`falsy-operator` can use since `patch 8.2.1794`(maybe...).
Older version can not use, so use trinary operator.

rhysd commented:
@heavenshell Thank you for the updates.

> How about using g:lsp_ignorecase?
> I added g:lsp_ignorecase flag to use case sensitive or case insensitive in sort function.

That makes sense to me. However Vim already has the option `'ignorecase'`. We can also reuse it for the sorting function. Or is it be better to have a separate option dedicated for vim-lsp?
I have no strong opinion here. @prabirshrestha, what do you think?

rhysd commented on autoload/lsp/omni.vim:
```diff
@@ -264,6 +264,21 @@ function! s:get_completion_result(server_name, data) abort
+    let l:text1 = l:text1 ? l:text1 : a:i1['label']
+    let l:text2 = l:text2 ? l:text2 : a:i2['label']
+
+    if g:lsp_ignorecase
+        return tolower(l:text1) == tolower(l:text2) ? 0 : tolower(l:text1) > tolower(l:text2) ? 1 : -1
+    else
+        return l:text1 == l:text2 ? 0 : l:text1 > l:text2 ? 1 : -1
+    endif
```
```suggestion
    if g:lsp_ignorecase
        return l:text1 ==? l:text2 ? 0 : l:text1 >? l:text2 ? 1 : -1
    else
        return l:text1 ==# l:text2 ? 0 : l:text1 ># l:text2 ? 1 : -1
    endif
```

Vim has case-sensitive/case-insensitive string comparison operators as builtin `==#`, `==?`, `>#`, `>?`.

rhysd commented on doc/vim-lsp.txt:
```diff
@@ -1202,6 +1203,16 @@ only possible by adding to the already typed prefix (even if you're using the
 
+sort                                              *vim-lsp-completion-sort*
+
+You can sort the completion items returned from the server by using the `sort` key in the server info's `config` |dict|.
```
This line is too long. Please take care about line width. Other lines are also exceeding the width imit.

heavenshell replied:
Thank you for pointed out!
Fixed!

rhysd commented on doc/vim-lsp.txt:
```diff
@@ -1202,6 +1203,16 @@ only possible by adding to the already typed prefix (even if you're using the
+
+You can sort the completion items returned from the server by using the `sort` key in the server info's `config` |dict|.
+The value of the `sort` key is itself a |dict| containing at least a key
+`max`, which specifies max number of completion items count before ignore sort for performance reason.
```
nit:

```suggestion
`max`, which specifies max number of completion items count before giving up sorting for performance reason.
```

rhysd commented on autoload/lsp/omni.vim:
```diff
@@ -288,6 +303,13 @@ function! lsp#omni#get_vim_completion_items(options) abort
         let l:incomplete = 0
     endif
 
+    let l:sort = has_key(l:server, 'config') && has_key(l:server['config'], 'sort') ? l:server['config']['sort'] : v:none
```
`v:none` does not exist on Neovim. `v:null` is available instead.

heavenshell replied:
Oh.. I did not know about this.
Use `v:null` instead!

prabirshrestha commented:
We should use the vim setting for ignorecase.

rhysd commented:
> We should use the vim setting for ignorecase.

Yes, `g:lsp_ignorecase` inherits Vim's `'ignorecase'` option. By default, the behavior is the same as referring the option. When a user wants to customize sensitivity only for vim-lsp, it is also possible.

heavenshell commented:
Thank you for review!
I Applyed suggested changes(Thank you so much!) and fixed points that were pointed out.

Please take another look 🙏

rhysd commented:
Neovim nightly build is now broken. We can ignore it.

https://github.com/neovim/neovim/releases/nightly
